# Programming with Python

#### Module-1 Graded Project-1

# Part - 1

##### 1. Answer the following questions.

    a. Given the string 'great', use an appropriate index command that returns from the string. Given s = 'great'.

In [1]:
s = 'great'
s[3]

'a'

    b. Write a program to find out the second smallest number in the list given below.
    
    li=[32, 3, 5, 8, 33, 7]

In [2]:
li = [32, 3, 5, 8, 33, 7]
li.sort()
print("The second smallest number from the list is ")
print(li[1])

The second smallest number from the list is 
5


    c. Find out the unique values of the list given below
    mylist=[1,1,5,3,3,6,11,11]

In [3]:
import numpy as np
mylist=[1,1,5,3,3,6,11,11]
print(np.unique(mylist))
# or
mylist_unique = set(mylist)
print(list(mylist_unique))

[ 1  3  5  6 11]
[1, 3, 5, 6, 11]


    d. Write a program to pring cube root of 216 uning appropriate syntax

In [4]:
x = 216
cube_root = (round(x**(1/3)))
print(cube_root)

6


##### 2. Replace the value 'hello' in the below nested list with 'bye'

a = [1, 2, [3, 4, 'hello']]

In [5]:
a = [1, 2, [3, 4, 'hello']]
print(a)
a[2][2] = 'bye'
print("After replacing 'hello' with 'bye'")
print(a)

[1, 2, [3, 4, 'hello']]
After replacing 'hello' with 'bye'
[1, 2, [3, 4, 'bye']]


##### 3. Given the below dictionary, write a program that returns ['object0'].
##### dict = {'a':9, 'b':(2,3,1), 'c':['object0', 'object1','object2']}

In [6]:
dict1 = {'a':9, 'b':(2,3,1), 'c':['object0','object1','object2']}
dict1
print(dict1['c'][0])

object0


##### 4. Write a program to calculate the area of a circle. The program should ask for an appropriate input and print the computed are.


    a. Input text: Please enter the radius of the circle.
    b. If the user enters a value of 5. the output be: The are of the circle is 78.54
    c. This program should indicate that the input is invalid if the user enters a character instead of a number. For instance if the user enters a value of 'a' instead of 5 in the previous example the program should prompt: invalid input, please enter a number

In [7]:
radius = float(input("please enter the radius of the cirlce " ,))
if type(radius) == float:
    area_of_Cirlce = 3.14159 * (radius*radius)
    print("The are of the circle is ""%.2f" %area_of_Cirlce)
else:
    print("Invalid input, please enter a number")       

please enter the radius of the cirlce 5
The are of the circle is 78.54


##### 5. Write a program which will take two list X & Y as input and it will return a list having all the odd numbers present in the two lists. (Note:- common elements can be repeated)
##### Input list: X=[1,2,3,4,5] & Y=[5,6,7,8,9]

In [8]:
import pandas as pd
import numpy as np

X = [1,2,3,4,5]
Y = [5,6,7,8,9]
a = X + Y
i = 0
b = list(filter(lambda i:i%2 != 0, a))
b

[1, 3, 5, 5, 7, 9]

###### 6. Write a function that prints the integers from 1 to 15. But for multiples of three print "GREAT" instead of the number, and for the multiples of five print "LEARNING", For numbers which are multiples of both three and five print "GREATLEARNING"

In [9]:
for i in range(1,16):
    if i % 3 == 0 and i % 5 == 0:
        print("GREATLEARNING", sep="\n")
    elif i % 3 == 0:
        print("GREAT", sep="\n")
    elif i % 5 == 0:
        print("LEARNING", sep="\n")
    elif i % 3 == 0 and i % 5 == 0:
        print("GREATLEARNING", sep="\n")
    
    else:
        print(i)

1
2
GREAT
4
LEARNING
GREAT
7
8
GREAT
LEARNING
11
GREAT
13
14
GREATLEARNING


# Part 2

##### Dataset: Online_sales.csv
##### Domain: Retail
##### Objective: Analyzing sales data to understand sales trend

###### 1. Import necessary libraries and read the provided dataset (online_sales.csv) and check the top 5 and random 5 samples of the dataframe.

In [10]:
import pandas as pd
import numpy as numpy

In [11]:
retail = pd.read_csv(r'C:/Users/Varad Rajesh Morde/Documents/IIIT DELHI GREATLEARNING/PROGRAMMING WITH PYTHON/Python poject 1/online_sales.csv')

In [12]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
240002    558077    47590A          BLUE HAPPY BIRTHDAY BUNTING        12   
240003    558077     82600                NO SINGING METAL SIGN        12   
240004    558077    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
240005    558077     22746        POPPY'S PLAYHOUSE LIVINGROOM         12   
240006    558077     23251         VINTAGE RED ENAMEL TRIM MUG         12   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
240002  6/26/11 10:59       5.45     13408.0  United Kingdom  
240003  6/26/11 10:59       2.10     13408.0  United Kingdom  
240004  6/26/11 10:59       2.55     13408.0  United Kingdom  
240005  6/26/11 10:59       2.10     13408.0  United Kingdom  
240006  6/26/11 10:59       1.25     13408.0            Unit  

[240007 rows x 8 columns]

In [13]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom

In [14]:
retail.sample(5)

InvoiceNo StockCode                       Description  Quantity  \
67459     541830     20979     36 PENCILS TUBE RED RETROSPOT         1   
52623     540803     22194            BLACK DINER WALL CLOCK         2   
57458     541206     22897           OVEN MITT APPLES DESIGN        50   
30371     538862     22630              DOLLY GIRL LUNCH BOX         1   
138849    548214    84970L  SINGLE HEART ZINC T-LIGHT HOLDER        12   

           InvoiceDate  UnitPrice  CustomerID         Country  
67459    1/21/11 17:09       2.46         NaN  United Kingdom  
52623    1/11/11 12:30       8.50     14669.0  United Kingdom  
57458    1/14/11 12:24       1.25     14646.0     Netherlands  
30371   12/14/10 14:11       1.95     16710.0  United Kingdom  
138849    3/30/11 9:17       0.95     13720.0  United Kingdom

###### 2. Check info of the dataframe and write your obesrvations. Comment on data types and shape of the dataset

In [15]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240007 entries, 0 to 240006
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    240007 non-null  object 
 1   StockCode    240007 non-null  object 
 2   Description  239106 non-null  object 
 3   Quantity     240007 non-null  int64  
 4   InvoiceDate  240007 non-null  object 
 5   UnitPrice    240007 non-null  float64
 6   CustomerID   172782 non-null  float64
 7   Country      240007 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 14.6+ MB


As we can the above info which provides us total number of rows as RangeIndex which are 240007.

The above summary also includes list of all comlumns and list of total count of Non-null values present in each column.

Also we can see the datatypes in Dtype as there are (2 float64, 1 int64 and 5 object)data types in the dataset.

Shape of dataframe we can say it as dataframe has 240007 rows and 8 columns

###### 3. Check for null values and report the percentage of null values of each column. And drop the rows having null values in it.

In [16]:
retail_null_values = np.where(retail.isnull() == True)
retail_null_values

(array([   622,    622,   1443, ..., 239905, 239906, 239907], dtype=int64),
 array([2, 6, 6, ..., 6, 6, 6], dtype=int64))

In [17]:
percent_null = retail.isnull().sum() * 100 / len(retail.index)

In [18]:
percent_null     #COLUMN WISE PERCENTAGE OF NULL VALUES

InvoiceNo       0.000000
StockCode       0.000000
Description     0.375406
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     28.009600
Country         0.000000
dtype: float64

In [19]:
retail_null_data = pd.DataFrame({'Col' : retail.columns, 'percentage_null' : percent_null})

In [20]:
retail_null_data    #df for column wise percentage of null values

Col  percentage_null
InvoiceNo      InvoiceNo         0.000000
StockCode      StockCode         0.000000
Description  Description         0.375406
Quantity        Quantity         0.000000
InvoiceDate  InvoiceDate         0.000000
UnitPrice      UnitPrice         0.000000
CustomerID    CustomerID        28.009600
Country          Country         0.000000

##### To drop the rows containing null values ==>

In [21]:
null_set = pd.DataFrame(retail)

In [22]:
null_set.dropna()

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
240002    558077    47590A          BLUE HAPPY BIRTHDAY BUNTING        12   
240003    558077     82600                NO SINGING METAL SIGN        12   
240004    558077    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
240005    558077     22746        POPPY'S PLAYHOUSE LIVINGROOM         12   
240006    558077     23251         VINTAGE RED ENAMEL TRIM MUG         12   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
240002  6/26/11 10:59       5.45     13408.0  United Kingdom  
240003  6/26/11 10:59       2.10     13408.0  United Kingdom  
240004  6/26/11 10:59       2.55     13408.0  United Kingdom  
240005  6/26/11 10:59       2.10     13408.0  United Kingdom  
240006  6/26/11 10:59       1.25     13408.0            Unit  

[172782 rows x 8 columns]

###### 4. Check the statistical summary of the dataset.

In [23]:
retail.describe(include = 'all')

InvoiceNo StockCode                         Description       Quantity  \
count     240007    240007                              239106  240007.000000   
unique     12468      3645                                3617            NaN   
top       537434    85123A  WHITE HANGING HEART T-LIGHT HOLDER            NaN   
freq         675      1294                                1319            NaN   
mean         NaN       NaN                                 NaN       9.277646   
std          NaN       NaN                                 NaN     223.061608   
min          NaN       NaN                                 NaN  -74215.000000   
25%          NaN       NaN                                 NaN       1.000000   
50%          NaN       NaN                                 NaN       3.000000   
75%          NaN       NaN                                 NaN      10.000000   
max          NaN       NaN                                 NaN   74215.000000   

          InvoiceDate      UnitPrice     CustomerID         Country  
count          240007  240007.000000  172782.000000          240007  
unique          11240            NaN            NaN              38  
top     12/6/10 16:57            NaN            NaN  United Kingdom  
freq              675            NaN            NaN          220279  
mean              NaN       5.124265   15274.819941             NaN  
std               NaN     119.992279    1725.093177             NaN  
min               NaN       0.000000   12346.000000             NaN  
25%               NaN       1.250000   13842.000000             NaN  
50%               NaN       2.100000   15132.000000             NaN  
75%               NaN       4.210000   16814.000000             NaN  
max               NaN   38970.000000   18287.000000             NaN

###### 5. Drop the instances having quantity less than zero.(Note:- refer to the 'Quantity' feature)

In [24]:
np.where(retail['Quantity'] < 0)

(array([   141,    154,    235, ..., 239550, 239551, 239552], dtype=int64),)

In [25]:
retail['Quantity'][141]

-1

In [26]:
Quan_less_zero = retail[retail['Quantity'] < 0]

In [27]:
Quan_less_zero

InvoiceNo StockCode                          Description  Quantity  \
141      C536379         D                             Discount        -1   
154      C536383    35004C      SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556       PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984     PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983     PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                                  ...       ...   
239404   C558048         M                               Manual        -1   
239549   C558059     23243  SET OF TEA COFFEE SUGAR TINS PANTRY        -1   
239550   C558059     82600                NO SINGING METAL SIGN        -2   
239551   C558060     22988                    SOLDIERS EGG CUP         -1   
239552   C558061     23245           SET OF 3 REGENCY CAKE TINS        -1   

          InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/10 9:41      27.50     14527.0  United Kingdom  
154      12/1/10 9:49       4.65     15311.0  United Kingdom  
235     12/1/10 10:24       1.65     17548.0  United Kingdom  
236     12/1/10 10:24       0.29     17548.0  United Kingdom  
237     12/1/10 10:24       0.29     17548.0  United Kingdom  
...               ...        ...         ...             ...  
239404  6/24/11 13:52     542.20         NaN            EIRE  
239549  6/24/11 15:03       4.95     14194.0  United Kingdom  
239550  6/24/11 15:03       1.85     14194.0  United Kingdom  
239551  6/24/11 15:05       1.06     17809.0  United Kingdom  
239552  6/24/11 15:15       4.95     12809.0        Portugal  

[5087 rows x 8 columns]

In [28]:
Quan_less_zero_index = retail[retail['Quantity'] < 0].index

In [29]:
less_than_zero = pd.DataFrame(retail)

In [30]:
less_than_zero.drop(Quan_less_zero_index)     #here we are now droping the instances which are having quantity less than zero

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
240002    558077    47590A          BLUE HAPPY BIRTHDAY BUNTING        12   
240003    558077     82600                NO SINGING METAL SIGN        12   
240004    558077    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
240005    558077     22746        POPPY'S PLAYHOUSE LIVINGROOM         12   
240006    558077     23251         VINTAGE RED ENAMEL TRIM MUG         12   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
240002  6/26/11 10:59       5.45     13408.0  United Kingdom  
240003  6/26/11 10:59       2.10     13408.0  United Kingdom  
240004  6/26/11 10:59       2.55     13408.0  United Kingdom  
240005  6/26/11 10:59       2.10     13408.0  United Kingdom  
240006  6/26/11 10:59       1.25     13408.0            Unit  

[234920 rows x 8 columns]

###### 6. Check unique values of the country and report the name of the country that has the highest number of instances/rows.

In [31]:
unique_val = retail['Country'].unique()

In [32]:
pd.DataFrame(unique_val)

0
0         United Kingdom
1                 France
2              Australia
3            Netherlands
4                Germany
5                 Norway
6                   EIRE
7            Switzerland
8                  Spain
9                 Poland
10              Portugal
11                 Italy
12               Belgium
13             Lithuania
14                 Japan
15               Iceland
16       Channel Islands
17               Denmark
18                Cyprus
19                Sweden
20               Austria
21                Israel
22               Finland
23               Bahrain
24                Greece
25             Hong Kong
26             Singapore
27               Lebanon
28  United Arab Emirates
29          Saudi Arabia
30        Czech Republic
31                Canada
32           Unspecified
33                Brazil
34                   USA
35    European Community
36                 Malta
37                  Unit

###### name the country that has the highest number of instances/rows

In [33]:
high_num_occor = pd.DataFrame(retail)

In [34]:
high_num_occor['Country'].max

<bound method NDFrame._add_numeric_operations.<locals>.max of 0         United Kingdom
1         United Kingdom
2         United Kingdom
3         United Kingdom
4         United Kingdom
               ...      
240002    United Kingdom
240003    United Kingdom
240004    United Kingdom
240005    United Kingdom
240006              Unit
Name: Country, Length: 240007, dtype: object>

In [35]:
#The above shows the United Kingdom is the country that has highest number of instances/rows 

###### 7. Create a new column with the name as 'sales' having total sales. The total sales is defined as Quantity*UnitPrice.

In [36]:
retail['sales'] = retail['Quantity'] * retail['UnitPrice']

In [37]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
240002    558077    47590A          BLUE HAPPY BIRTHDAY BUNTING        12   
240003    558077     82600                NO SINGING METAL SIGN        12   
240004    558077    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
240005    558077     22746        POPPY'S PLAYHOUSE LIVINGROOM         12   
240006    558077     23251         VINTAGE RED ENAMEL TRIM MUG         12   

          InvoiceDate  UnitPrice  CustomerID         Country  sales  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  15.30  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  22.00  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
...               ...        ...         ...             ...    ...  
240002  6/26/11 10:59       5.45     13408.0  United Kingdom  65.40  
240003  6/26/11 10:59       2.10     13408.0  United Kingdom  25.20  
240004  6/26/11 10:59       2.55     13408.0  United Kingdom  81.60  
240005  6/26/11 10:59       2.10     13408.0  United Kingdom  25.20  
240006  6/26/11 10:59       1.25     13408.0            Unit  15.00  

[240007 rows x 9 columns]

###### 8. Report the top 5 countries in terms of sales.

###### a. Consider the size of sales
###### b. Consider the mean value of sales

In [38]:
max(retail['sales'])

77183.6

In [39]:
retail.iloc[retail['sales'].nlargest(n=5)]

InvoiceNo StockCode                        Description  Quantity  \
77183    542713     84991        60 TEATIME FAIRY CAKE CASES        24   
38970    539607     21240                  BLUE POLKADOT CUP         8   
13541    537434     22322     BIRD DECORATION GREEN POLKADOT         1   
8142     537126     22543           MINI JIGSAW BAKE A CAKE          1   
6539     536947     22072  RED RETROSPOT TEA CUP AND SAUCER          4   

          InvoiceDate  UnitPrice  CustomerID         Country  sales  
77183   1/31/11 13:27       0.55     12668.0         Germany  13.20  
38970  12/20/10 14:10       0.85     12681.0          France   6.80  
13541   12/6/10 16:57       1.66         NaN  United Kingdom   1.66  
8142    12/5/10 12:13       0.42     18118.0  United Kingdom   0.42  
6539    12/3/10 12:29       3.75     13013.0  United Kingdom  15.00

In [40]:
retail.iloc[retail['sales'].nlargest(n=5)]['Country']

77183           Germany
38970            France
13541    United Kingdom
8142     United Kingdom
6539     United Kingdom
Name: Country, dtype: object

In [41]:
top_five = pd.DataFrame(retail.iloc[retail['sales'].nlargest(n=5)]['Country'])

In [42]:
top_five

Country
77183         Germany
38970          France
13541  United Kingdom
8142   United Kingdom
6539   United Kingdom

the above gives us the top 5 countries in terms of sales

In [43]:
retail.groupby(by='Country')['sales'].agg([np.size]).sort_values(by = 'size', ascending = False).head()

size
Country               
United Kingdom  220279
Germany           4208
France            3642
EIRE              3034
Spain             1142

In [44]:
retail.groupby(by='Country')['sales'].agg([np.mean]).sort_values(by = 'mean', ascending=False).head()

mean
Country                
Australia    123.164003
Netherlands  110.088713
Japan         84.194183
Sweden        77.243900
Denmark       51.285652

In [45]:
#top_five.mean values of sales is above

###### 9. Report the top 5 products which bring the highest sales. Use StockCode product information.

In [46]:
retail.iloc[retail['sales'].nlargest(5)]

InvoiceNo StockCode                        Description  Quantity  \
77183    542713     84991        60 TEATIME FAIRY CAKE CASES        24   
38970    539607     21240                  BLUE POLKADOT CUP         8   
13541    537434     22322     BIRD DECORATION GREEN POLKADOT         1   
8142     537126     22543           MINI JIGSAW BAKE A CAKE          1   
6539     536947     22072  RED RETROSPOT TEA CUP AND SAUCER          4   

          InvoiceDate  UnitPrice  CustomerID         Country  sales  
77183   1/31/11 13:27       0.55     12668.0         Germany  13.20  
38970  12/20/10 14:10       0.85     12681.0          France   6.80  
13541   12/6/10 16:57       1.66         NaN  United Kingdom   1.66  
8142    12/5/10 12:13       0.42     18118.0  United Kingdom   0.42  
6539    12/3/10 12:29       3.75     13013.0  United Kingdom  15.00

In [47]:
retail.iloc[retail['sales'].nlargest(5)]['StockCode']

77183    84991
38970    21240
13541    22322
8142     22543
6539     22072
Name: StockCode, dtype: object

In [48]:
top_five_prod = pd.DataFrame(retail.iloc[retail['sales'].nlargest(5)]['StockCode'])

In [49]:
top_five_prod

StockCode
77183     84991
38970     21240
13541     22322
8142      22543
6539      22072

###### 10. Convert the 'InvoiceData' into a date format and report the month on which the maximum sales occur?

In [50]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240007 entries, 0 to 240006
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    240007 non-null  object 
 1   StockCode    240007 non-null  object 
 2   Description  239106 non-null  object 
 3   Quantity     240007 non-null  int64  
 4   InvoiceDate  240007 non-null  object 
 5   UnitPrice    240007 non-null  float64
 6   CustomerID   172782 non-null  float64
 7   Country      240007 non-null  object 
 8   sales        240007 non-null  float64
dtypes: float64(3), int64(1), object(5)
memory usage: 16.5+ MB


In [51]:
retail['InvoiceDate'] = pd.to_datetime(retail['InvoiceDate'])

In [52]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240007 entries, 0 to 240006
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    240007 non-null  object        
 1   StockCode    240007 non-null  object        
 2   Description  239106 non-null  object        
 3   Quantity     240007 non-null  int64         
 4   InvoiceDate  240007 non-null  datetime64[ns]
 5   UnitPrice    240007 non-null  float64       
 6   CustomerID   172782 non-null  float64       
 7   Country      240007 non-null  object        
 8   sales        240007 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 16.5+ MB


In [53]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
240002    558077    47590A          BLUE HAPPY BIRTHDAY BUNTING        12   
240003    558077     82600                NO SINGING METAL SIGN        12   
240004    558077    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
240005    558077     22746        POPPY'S PLAYHOUSE LIVINGROOM         12   
240006    558077     23251         VINTAGE RED ENAMEL TRIM MUG         12   

               InvoiceDate  UnitPrice  CustomerID         Country  sales  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  15.30  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  22.00  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
...                    ...        ...         ...             ...    ...  
240002 2011-06-26 10:59:00       5.45     13408.0  United Kingdom  65.40  
240003 2011-06-26 10:59:00       2.10     13408.0  United Kingdom  25.20  
240004 2011-06-26 10:59:00       2.55     13408.0  United Kingdom  81.60  
240005 2011-06-26 10:59:00       2.10     13408.0  United Kingdom  25.20  
240006 2011-06-26 10:59:00       1.25     13408.0            Unit  15.00  

[240007 rows x 9 columns]

In [54]:
retail['month'] = retail['InvoiceDate'].dt.month

In [55]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
240002    558077    47590A          BLUE HAPPY BIRTHDAY BUNTING        12   
240003    558077     82600                NO SINGING METAL SIGN        12   
240004    558077    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
240005    558077     22746        POPPY'S PLAYHOUSE LIVINGROOM         12   
240006    558077     23251         VINTAGE RED ENAMEL TRIM MUG         12   

               InvoiceDate  UnitPrice  CustomerID         Country  sales  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  15.30   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  22.00   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34   
...                    ...        ...         ...             ...    ...   
240002 2011-06-26 10:59:00       5.45     13408.0  United Kingdom  65.40   
240003 2011-06-26 10:59:00       2.10     13408.0  United Kingdom  25.20   
240004 2011-06-26 10:59:00       2.55     13408.0  United Kingdom  81.60   
240005 2011-06-26 10:59:00       2.10     13408.0  United Kingdom  25.20   
240006 2011-06-26 10:59:00       1.25     13408.0            Unit  15.00   

        month  
0          12  
1          12  
2          12  
3          12  
4          12  
...       ...  
240002      6  
240003      6  
240004      6  
240005      6  
240006      6  

[240007 rows x 10 columns]

In [61]:
retail.iloc[retail['sales'].nlargest(n=1)]['month']

77183    1
Name: month, dtype: int64

In [59]:
retail.groupby(by = 'month').sales.max()

month
1     77183.60
2      3215.52
3      2583.76
4      6539.40
5      8142.75
6     38970.00
12    13541.33
Name: sales, dtype: float64